src/backend/drm/renderer/mod.rs

```diff
@@ -0,0 +1,2309 @@
+
+    fn add_framebuffer<D: drm::control::Device>(
+        &self,
+        drm: &D,
+        buffer: ExportBuffer<'_, B>,
+    ) -> Result<Option<Self::Framebuffer>, Self::Error> {
+        let guard = self.lock().unwrap();
+        guard.add_framebuffer(drm, buffer)
+    }
+}
+
+impl<F, B> ExportFramebuffer<B> for Rc<RefCell<F>>
+where
+    F: ExportFramebuffer<B>,
+    B: Buffer,
+{
+    type Framebuffer = <F as ExportFramebuffer<B>>::Framebuffer;
+    type Error = <F as ExportFramebuffer<B>>::Error;
+
+    fn add_framebuffer<D: drm::control::Device>(
+        &self,
+        drm: &D,
+        buffer: ExportBuffer<'_, B>,
+    ) -> Result<Option<Self::Framebuffer>, Self::Error> {
+        self.borrow().add_framebuffer(drm, buffer)
+    }
+}
+
+/// Specialized render for drm output which uses planes
+#[derive(Debug)]
+pub struct DrmRenderer<A, B, D, F, U>
+where
+    A: Allocator<B>,
+    B: Buffer,
+    D: AsRawFd + 'static,
+    F: ExportFramebuffer<B>,
+    <F as ExportFramebuffer<B>>::Framebuffer: std::fmt::Debug + 'static,
+{
+    output: Output,
+    surface: Arc<DrmSurface<D>>,
+    planes: Planes,
+    damage_tracked_renderer: DamageTrackedRenderer,
+
+    framebuffer_exporter: F,
+
+    current_frame: FrameState<D, DrmScanoutBuffer<D, B, <F as ExportFramebuffer<B>>::Framebuffer>>,
+    #[allow(clippy::type_complexity)]
+    pending_frame: Option<(
+        FrameState<D, DrmScanoutBuffer<D, B, <F as ExportFramebuffer<B>>::Framebuffer>>,
+        U,
+    )>,
+    #[allow(clippy::type_complexity)]
+    queued_frame: Option<(
+        FrameState<D, DrmScanoutBuffer<D, B, <F as ExportFramebuffer<B>>::Framebuffer>>,
+        U,
+    )>,
+    next_frame: Option<FrameState<D, DrmScanoutBuffer<D, B, <F as ExportFramebuffer<B>>::Framebuffer>>>,
+
+    swapchain: Swapchain<A, B>,
+
+    cursor_size: Size<i32, Physical>,
+    cursor_swapchain: Swapchain<A, B>,
+    cursor_damage_tracked_renderer: DamageTrackedRenderer,
+
+    last_scale: Scale<f64>,
+
+    element_states: IndexMap<Id, ElementState<D, <F as ExportFramebuffer<B>>::Framebuffer>>,
+
+    debug_flags: DebugFlags,
+}
+
+const SUPPORTED_FORMATS: &[DrmFourcc] = &[DrmFourcc::Argb8888, DrmFourcc::Xrgb8888];
+
+impl<A, B, D, F, U> DrmRenderer<A, B, D, F, U>
+where
+    A: Allocator<B> + Clone,
+    <A as Allocator<B>>::Error: std::error::Error + Send + Sync,
+    B: Buffer + AsDmabuf,
+    <B as AsDmabuf>::Error: std::error::Error + Send + Sync + std::fmt::Debug,
+    D: AsRawFd + std::fmt::Debug + 'static,
+    F: ExportFramebuffer<B>,
+    <F as ExportFramebuffer<B>>::Framebuffer: std::fmt::Debug + 'static,
+    <F as ExportFramebuffer<B>>::Error: std::error::Error + Send + Sync,
+{
+    /// Initialize a new [`DrmRenderer`]
+    pub fn new<L>(
+        output: &Output,
+        surface: DrmSurface<D>,
+        mut allocator: A,
+        framebuffer_exporter: F,
+        renderer_formats: HashSet<DrmFormat>,
+        cursor_size: Size<u32, BufferCoords>,
```

Comment:
I think cursor_size should be an option because input/cursor isn't required for all compositors.

Reply:
The drm renderer currently tries to assign whatever fits on the cursor plane, so technically it is not required to be a cursor element. With the latest changes the cursor plane can be completely disabled by not passing it to the constructor.